dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -766,12 +766,50 @@ void CudaIcoCodeGen::generateAllAPIRunFunctions(
     chains.insert(chainCollector.getChains().begin(), chainCollector.getChains().end());
   }
 
+  const std::string wrapperName = stencilInstantiation->getName();
+  {
+    std::vector<std::stringstream> apiRunFunStreams(fromHost ? 2 : 1);
+    if(stencils.empty()) {
+      std::vector<std::unique_ptr<MemberFunction>> apiRunFuns;
+      if(fromHost) {
+        apiRunFuns.push_back(
+            std::make_unique<MemberFunction>("double", "run_" + wrapperName + "_from_c_host",
+                                             apiRunFunStreams[0], /*indent level*/ 0, onlyDecl));
+        apiRunFuns.push_back(
+            std::make_unique<MemberFunction>("double", "run_" + wrapperName + "_from_fort_host",
+                                             apiRunFunStreams[1], /*indent level*/ 0, onlyDecl));
+      } else {
+        apiRunFuns.push_back(std::make_unique<MemberFunction>(
+            "double", "run_" + wrapperName, apiRunFunStreams[0], /*indent level*/ 0, onlyDecl));
+      }
+
+      for(auto& apiRunFun : apiRunFuns) {
+        apiRunFun->addArg("dawn::GlobalGpuTriMesh *mesh");
+        apiRunFun->addArg("int k_size");
+        for(auto accessID : stencilInstantiation->getMetaData().getAPIFields()) {
+          apiRunFun->addArg("::dawn::float_type *" +
+                            stencilInstantiation->getMetaData().getNameFromAccessID(accessID));
+        }
+        apiRunFun->finishArgs();
+        if(!onlyDecl) {
+          apiRunFun->startBody();
+          apiRunFun->addStatement("return 0.");
+        }
+        apiRunFun->commit();
+      }
+
+      for(const auto& stream : apiRunFunStreams) {
+        ssSW << stream.str() << ";\n";
+      }
+      return;
+    }
+  }
+
```

Comment:
Looks like it's almost a code clone of the non-empty stencil case, have you considered making the empty/non-empty distinction below? Maybe adding at line 875 something like
```
if(stencils.empty()) {
   apiRunFun->startBody();
   apiRunFun->addStatement("return 0.");
}
```
and letting the rest ( https://github.com/MeteoSwiss-APN/dawn/blob/fda39003acb1287624b12b49dcc9784bc5326f2e/dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp#L876-L900 ) be in the else block.

Reply:
I first wanted to avoid that, because the lines you are indicating are already inside a loop over the stencils (which is pointless of course since there is only one). I initially thought that the refactoring due to accesses to `stencil` would be quite involved, but looking at it again it turned out easier than expected.